src/satosa/backends/saml2.py

```diff
@@ -327,16 +327,17 @@ def authn_response(self, context, binding):
             del self.outstanding_queries[req_id]
 
         # check if the relay_state matches the cookie state
-        if context.state[self.name]["relay_state"] != context.request["RelayState"]:
-            satosa_logging(logger, logging.DEBUG,
-                           "State did not match relay state for state", context.state)
-            raise SATOSAAuthenticationError(context.state, "State did not match relay state")
+        if self.name in context.state and "relay_state" in context.state[self.name]:
```

Comment:
I would suggest `context.state.get(self.name, {}).get("relay_state") != context.request["RelayState"]` to avoid nesting if statements if possible (without hurting readability). It also avoids having to lookup 4 times (vs 2) in the context.state (2 in your first `if` and another 2 in the second `if`)

Reply:
I do not follow the thinking that nested expressions are better than nested if statements. 

I agree in general that using get() is simpler that a verbose 2-part condition. But in this case I think that separating the test for the existence of attribute&key from the comparison of the value does make the intention of this pice of code very explicit.

Reply:
Since the `context.state.pop(self.name, None)` works outside of the if statement and there is no other action on the `else` part, you could also have this written as one big `if` statement
The point of nested expressions vs multiple `if` statements is that they reduce the cyclomatic complexity by having fewer control flows to follow and to remember while walking down the `if` tree

My comment was wrong though in that it made it required (while your diff made it optional). So the correct proposal would be `if context.state.get(self.name, {}).get("relay_state") not in (None, context.request["RelayState"]):`

I think that since there is a check about `relay_state` existence and a comparison with `context.request["RelayState"]` in addition to the actions below (the `exception`) it is as explicit as it gets in either case

Reply:
Please feel free to modify the PR.

However, I am still not persuaded that concatenated functions (the Camel Pattern according to Robert C. Martin in his book "Clean Code") has better readability that nested ifs.

Reply:
I'm not going to modify anything since we don't agree

_sidenote_: I haven't heard of the Caravan Pattern even though I own and have read the book of uncle bob. Tried to look it up but end up with no results. Is it on a newer version maybe?